File: CPP/knapsack.cpp
#include <bits/stdc++.h>
using namespace std;

int knapsack(int wt[], int value[] ,int arr_size, int weight) 
{
    //function which returns the maximum cost of the knapsack
    int dp[arr_size+1][weight+1];
    
    //initialising dp array
    for(int i=0; i<arr_size+1; i++){
        for(int j=0; j<weight+1; j++){
            if(i==0 || j==0){
                dp[i][j]=0;
            }
        }
    }

    //filling the dp array
    for(int i=1; i<arr_size+1; i++){
        for(int j=1; j<weight+1; j++){
            if(wt[i-1]<=weight){
                dp[i][j]=max(dp[i-1][j], value[i-1]+dp[i-1][j-wt[i-1]]);
            }
            else{
                dp[i][j]=dp[i-1][j];
            }
        }
    }
    return dp[arr_size][weight];
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int arr_size; //array size
    cin>>arr_size;
    int wt[arr_size]; //weight array 
    int value[arr_size]; //value array of each weight
    for(int i=0; i<arr_size; i++){
        cin>>wt[i];
    }
    for(int i=0; i<arr_size; i++){
        cin>>value[i];
    }
    int weight; //total weight of knapsack
    cin>>weight;
    cout<<knapsack(wt, value, arr_size, weight)<<endl;
    return 0;
}

/*
0/1 knapsack implementation in C++
Input: Size of array , Array of weight, Array of value for each weight, total weight of knapsack
Output: Maximum Value/Price of knapsack

----------------------------------------
Sample Input1:
3
[10, 20, 30]
[60, 100, 120]
50
Output:
220

----------------------------------------

Time Complexity:
Worst and Average case = O(arr_size * weight)

Space Complexity: O(arr_size * weight)

*/

File: Java/Knapsack.java
import java.util.*;
public class Knapsack
{
    int solveKnapsack(int[] value, int[] weights, int capacity)
    {
        // basic checks
      if (capacity <= 0 || value.length == 0 || weights.length != value.length)
        return 0;
  
      int n = value.length;
      int[][] dp = new int[n][capacity + 1];
  
      // populate the capacity=0 columns, with '0' capacity we have '0' value
      for(int i=0; i < n; i++)
        dp[i][0] = 0;
  
      // if we have only one weight, we will take it if it is not more than the capacity
      for(int c=0; c <= capacity; c++) {
        if(weights[0] <= c)
          dp[0][c] = value[0];
      }
  
      // process all sub-arrays for all the capacities
      for(int i=1; i < n; i++)
      {
        for(int c=1; c <= capacity; c++)
        {
          int value1= 0, value2 = 0;
          // include the item, if it is not more than the capacity
          if(weights[i] <= c)
            value1 = value[i] + dp[i-1][c-weights[i]];
          // exclude the item
          value2 = dp[i-1][c];
          // take maximum
          dp[i][c] = Math.max(value1, value2);
        }
      }
      // maximum profit will be at the bottom-right corner.
      return dp[n-1][capacity];
    }
  
    public static void main(String[] args)
    {
      Scanner sc = new Scanner( System.in);
      Knapsack obj = new Knapsack();
      System.out.println("Enter the no of values");
      int n = sc.nextInt();
      int[] value = new int[n];
      int[] weights = new int[n];
      System.out.println("Enter the values");
      for(int i = 0; i<n; i++)
      value[i] = sc.nextInt();
      System.out.println("Enter the weights");
      for(int i = 0; i<n; i++)
      weights[i] = sc.nextInt();
      System.out.println("Enter the knapsack capacity");
      int w = sc.nextInt();
      int maxvalue = obj.solveKnapsack(value, weights, w);
      System.out.println("Total knapsack value ---> " + maxvalue);
    }//end of main
  }//end of class
/*
Sample Input & Output -
INPUT:
Enter the no of values
3
Enter the values
50 100 120
Enter the weights
10 20 30

OUTPUT:
Total knapsack value ---> 220
-----------------------------
TIME COMPLEXITY:
O(N*C)
[(where ‘N’ is the number of items and ‘C’ is the knapsack capacity)]
---------------------------------------------------------------------
SPACE COMPLEXITY:
O(N∗C)
*/